### PR TITLE
cp: add support for -x/--one-file-system

### PR DESCRIPTION
**Pull request by nomius10**
#1762 

Luckily, WalkDir provides [an option](https://docs.rs/walkdir/2.3.1/walkdir/struct.WalkDir.html?search=#method.same_file_system) for avoiding the traversal of filesystem boundaries. This functionality is only supported for [Unix and Windows](https://docs.rs/walkdir/2.3.1/src/walkdir/util.rs.html#5) at the moment.

I've run an experiment locally, and the behavior looks correct:
```
├── dst
│   ├── do_copy
│   │   └── do_copy.txt
│   ├── do_copy.txt
│   └── mount
├── source
│   ├── do_copy
│   │   └── do_copy.txt
│   ├── do_copy.txt
│   └── mount
│       ├── DONT_copy
│       │   └── DONT_copy.txt
│       ├── DONT_copy.txt
│       └── lost+found [error opening dir]
```

### CODE REVIEW

File: src/uu/cp/src/cp.rs [hunk 1]
```diff
@@ -563,6 +563,7 @@ impl Options {
         let not_implemented_opts = vec![
             OPT_COPY_CONTENTS,
             OPT_SPARSE,
+            #[cfg(not(any(windows, unix)))]
```
Comment (sylvestre) [inline src/uu/cp/src/cp.rs:566]: Unix implies Linux and Mac?

**Comment by sylvestre** [comment]
Even if I am aware it is hard to test correctly, please add a test.

At least that the option is present and taken into account
thanks

File: src/uu/cp/src/cp.rs (hunk 1 above)
Comment (nomius10) [reply src/uu/cp/src/cp.rs:566]: I'm not sure ATM. I've only reflected what [WalkDir supports](https://docs.rs/walkdir/2.3.1/src/walkdir/util.rs.html#19)

Comment (sylvestre) [reply src/uu/cp/src/cp.rs:566]: yeah, this is all OS pretty much :)
(except Fuschia)

**Review by sylvestre — CHANGES_REQUESTED** [review]
Needs tests!

**Comment by nomius10** [comment]
does the CI allow running commands with sudo? I need it to run `mount`

**Comment by sylvestre** [comment]
the ci is running as root by default 
See
https://github.community/t/how-to-run-action-as-a-non-root-user/17572/2

**Comment by nomius10** [comment]
Thanks! I'll look tomorrow into implementing a test that uses `mount`.
Another thought that crossed my mind is that `--link` should imply `--one-file-system`, if that doesn't already happen.

**Comment by nomius10** [comment]
I've just realized I could eliminate a lot of boilerplate. I'll refactor the code

File: Cargo.toml [hunk 2]
```diff
@@ -347,6 +347,7 @@ tempfile = "= 3.1.0"
 time = "0.1"
 unindent = "0.1"
 uucore = { version=">=0.0.7", package="uucore", path="src/uucore", features=["entries"] }
+walkdir = "2.2"
```
Comment (sylvestre) [inline Cargo.toml:350]: I don't think it should be global.
Instead, it should src/uu/cp/Cargo.toml

File: tests/by-util/test_cp.rs [hunk 3]
```diff
@@ -31,6 +31,9 @@ static TEST_COPY_FROM_FOLDER: &str = "hello_dir_with_file/";
 static TEST_COPY_FROM_FOLDER_FILE: &str = "hello_dir_with_file/hello_world.txt";
 static TEST_COPY_TO_FOLDER_NEW: &str = "hello_dir_new";
 static TEST_COPY_TO_FOLDER_NEW_FILE: &str = "hello_dir_new/hello_world.txt";
+static TEST_MOUNT_COPY_FROM_FOLDER: &str = "dir_with_mount";
```
Comment (sylvestre) [inline tests/by-util/test_cp.rs:34]: could you please fix this error ?

File: tests/by-util/test_cp.rs [hunk 4]
```diff
@@ -1001,3 +1004,67 @@ fn test_cp_target_file_dev_null() {
 
     assert!(at.file_exists(file2));
 }
+
+#[test]
```
Comment (sylvestre) [inline tests/by-util/test_cp.rs:1008]: please rustfmt the file

**Review by sylvestre — CHANGES_REQUESTED** [review]
Looks great, just some minor things

**Review by nomius10 — COMMENTED** [review]
An observation: from many trial-and-error commits, I've found that the Ubuntu/FreeBSD CI containers don't run as root. Therefore, nothing actually gets tested... Locally it does works for me in WSL2.

Having so many commits is a bit embarrassing. Could you squash them from github when you make the merge?

File: Cargo.toml (hunk 2 above)
Comment (sylvestre) [reply Cargo.toml:350]: oh, ok, sorry

**Comment by sylvestre** [comment]
> Having so many commits is a bit embarrassing. 

Don't worry, this is fine :)

**Pull request merged**